pkg/querier/queryrange/log_result_cache.go

```diff
@@ -0,0 +1,280 @@
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error storing cache", "err", err)
+	}
+	return resp, nil
+}
+
+func (l *logResultCache) handleHit(ctx context.Context, cacheKey string, cachedResquest *LokiRequest, lokiReq *LokiRequest) (queryrangebase.Response, error) {
+	l.metrics.CacheHit.Inc()
+	// we start with an empty response
+	result := emptyResponse(cachedResquest)
+	// if the request is the same and covers the whole time range,
+	// we can just return the cached result.
+	if lokiReq.GetStartTs().After(cachedResquest.GetStartTs()) || lokiReq.GetStartTs().Equal(cachedResquest.GetStartTs()) &&
+		lokiReq.GetEndTs().Before(cachedResquest.GetEndTs()) || lokiReq.GetEndTs().Equal(cachedResquest.GetEndTs()) {
+		return result, nil
+	}
+	// we could be missing data at the start and the end.
+	// so we're going to fetch what is missing.
+	var (
+		startRequest, endRequest *LokiRequest
+		startResp, endResp       *LokiResponse
+		updateCache              bool
+	)
+	g, ctx := errgroup.WithContext(ctx)
+
+	// if we're missing data at the start, start fetching from the start to the cached start.
+	if lokiReq.GetStartTs().Before(cachedResquest.GetStartTs()) {
+		g.Go(func() error {
+			startRequest = lokiReq.WithStartEndTime(lokiReq.GetStartTs(), cachedResquest.GetStartTs())
+			resp, err := l.next.Do(ctx, startRequest)
+			if err != nil {
+				return err
+			}
+			var ok bool
+			startResp, ok = resp.(*LokiResponse)
+			if !ok {
+				return fmt.Errorf("unexpected response type %T", resp)
+			}
+			return nil
+		})
+	}
+
+	// if we're missing data at the end, start fetching from the cached end to the end.
+	if lokiReq.GetEndTs().After(cachedResquest.GetEndTs()) {
+		g.Go(func() error {
+			endRequest = lokiReq.WithStartEndTime(cachedResquest.GetEndTs(), lokiReq.GetEndTs())
+			resp, err := l.next.Do(ctx, endRequest)
+			if err != nil {
+				return err
+			}
+			var ok bool
+			endResp, ok = resp.(*LokiResponse)
+			if !ok {
+				return fmt.Errorf("unexpected response type %T", resp)
+			}
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+
+	// if we have data at the start, we need to merge it with the cached data if it's empty and update the cache.
+	// If it's not empty only merge the response.
+	if startResp != nil {
+		if isEmpty(startResp) {
+			cachedResquest = cachedResquest.WithStartEndTime(startRequest.GetStartTs(), cachedResquest.GetEndTs())
+			updateCache = true
+		} else {
+			if startResp.Status != loghttp.QueryStatusSuccess {
+				return startResp, nil
+			}
+			result = mergeLokiResponse(startResp, result)
+		}
+	}
+
+	// if we have data at the end, we need to merge it with the cached data if it's empty and update the cache.
+	// If it's not empty only merge the response.
+	if endResp != nil {
+		if isEmpty(endResp) {
+			cachedResquest = cachedResquest.WithStartEndTime(cachedResquest.GetStartTs(), endRequest.GetEndTs())
```

Comment:
If we also have data on the start (line 224), this will override the start cached request, no?

Reply:
it is using the start of `cachedResquest` so whatever was set is reused here. Am I missing something ?

Reply:
Ah ok, got it. I got confused with the start and end times.

So if uncached data exists both at the start and at the end, the cache will be updated with time interval [startRequest.GetStartTs(), endRequest.GetEndTs()]. Couldn't this affect the interval part of the cache key? https://github.com/grafana/loki/blob/c5d27d0e8f696e6f29810f721b882eb8ee413f90/pkg/querier/queryrange/log_result_cache.go#L107 🤔

Reply:
It won't affect the key becaue it's forced to be aligned with the interval.

Reply:
But it is still possible that `startRequest.GetStartTs()` is before the aligned start interval and `endRequest.GetEndTs()` after the next start interval, right? As the example in the following diagram:
![cache_diagram](https://user-images.githubusercontent.com/5222392/156743205-9cd5f95b-2373-4898-8f9b-5d07f05596de.png)
The updated cache_n time interval is bigger than the expected size based on the split by duration interval.
With the current implementation, it is possible for this to happen, no?

The only issue here is that the updated cache_n could encompass one or more extra subsequent cache blocks, but that shouldn't be a problem.

Reply:
I don't think it's possible to be referring to different cache blocks at this point because we calculated `alignedStart` prior to seeing if we had data in the cache. Even if we only had a partial cache here, the range which we're checking will already align with the `alignedStart` cache key.

Reply:
I'm thinking about this. I don't think the start is a problem as it will be always either exactly at the start or within never before.

But the end can be problematic.

Reply:
In which case that seems to be a non-issue, the end might include more data.